### ISSUE TITLE
Reverse-mode adjoint of a pre-loop variable used after a dynamic (runtime-bound) loop is silently zero

### ISSUE THREAD

**Issue by Ramzee-S** [NVIDIA/warp#1694]
## Bug Description

In reverse-mode AD, the adjoint of a variable that is defined **before** a dynamic (runtime-bound) `for` loop and used **after** it is silently lost — `tape.backward()` returns 0 for its gradient. Variables read *inside* the loop and loop-carried accumulators get correct adjoints; only the pre-loop → post-loop path dies. With a static (compile-time) bound the same kernel differentiates correctly.

The forward pass is bit-identical in both variants, so nothing flags the corruption — we found it in a production FEM shell-element kernel where the through-thickness integration loop (`for k in range(n_thick)` with `n_thick` a kernel arg) sits between the element geometry computation (B-matrix, area, rotation) and the internal-force accumulation that consumes it. The entire geometry path of `d(force)/d(position)` vanished from the adjoint (~27× under-reported gradient) while every AD smoke test with loose tolerances kept passing. Same wrong values on CPU and CUDA, fp32 and fp64.

Possibly related: #1293 (nested dynamic loops, closed) and #1552 (`continue` in dynamic loops) — this one needs no nesting and no `continue`.

## Reproducer

```python
import warp as wp
wp.init()

@wp.kernel
def k_dyn(x: wp.array(dtype=float), s: wp.array(dtype=float), n: int,
          out: wp.array(dtype=float)):
    b = x[0] * 2.0            # pre-loop var derived from a differentiable input
    acc = float(0.0)
    for i in range(n):        # DYNAMIC bound (kernel arg)
        acc += s[i]
    out[0] = acc * b          # post-loop use of b

@wp.kernel
def k_sta(x: wp.array(dtype=float), s: wp.array(dtype=float), n: int,
          out: wp.array(dtype=float)):
    b = x[0] * 2.0
    acc = float(0.0)
    for i in range(3):        # STATIC bound
        acc += s[i]
    out[0] = acc * b

for kern, name in ((k_sta, "static"), (k_dyn, "dynamic")):
    x = wp.array([1.5], dtype=float, device="cpu", requires_grad=True)
    s = wp.array([1.0, 2.0, 3.0], dtype=float, device="cpu", requires_grad=True)
    out = wp.zeros(1, dtype=float, device="cpu", requires_grad=True)
    tape = wp.Tape()
    with tape:
        wp.launch(kern, dim=1, inputs=[x, s, 3], outputs=[out], device="cpu")
    tape.backward(out)
    print(f"{name}: d(out)/dx = {x.grad.numpy()[0]}  (expect 12.0)   "
          f"d(out)/ds[0] = {s.grad.numpy()[0]} (expect 3.0)")
```

Output:

```
static: d(out)/dx = 12.0  (expect 12.0)   d(out)/ds[0] = 3.0 (expect 3.0)
dynamic: d(out)/dx = 0.0  (expect 12.0)   d(out)/ds[0] = 3.0 (expect 3.0)
```

`d(out)/dx` should be `2 * (1+2+3) = 12` in both cases. Note `d(out)/ds` (the in-loop read) is correct even in the dynamic variant — only the pre-loop variable's post-loop contribution is dropped.

## Expected behavior

Identical gradients for static and dynamic loop bounds (or, if dynamic-loop adjoints are unsupported in some configuration, an error rather than silently wrong gradients).

## Workaround

Specialize the kernel per loop count via a factory closure (`for i in range(n)` with `n` a captured Python int) — static unrolling restores exact adjoints.

## Environment

- Warp 1.15.0 and 1.16.0.dev20260724 (both reproduce, CPU and CUDA devices)
- Linux 6.5.0-28-generic x86_64, glibc 2.35, Python 3.12.12
- GPU: NVIDIA RTX 4000 SFF Ada Generation, driver 550.144.03, CUDA toolkit 12.9

**Comment by shi-eric** [comment]
As documented in the [Dynamic Loops section of Warp's differentiability guide](https://nvidia.github.io/warp/stable/user_guide/differentiability.html#dynamic-loops), Warp currently does not replay or unroll dynamic loops during the backward pass. The result in this reproducer follows directly from that known limitation.

In this example, `b` itself is still available. Computing its adjoint for `out = acc * b` requires the forward value of `acc`. Since `acc` is computed by the dynamic loop and that loop is not replayed, `acc` has its initialized value of zero during replay. That makes the adjoint of `b`, and therefore `d(out)/dx`, zero. The `d(out)/ds` path still works because it depends on `b`, which was computed before the loop and is available during replay. The static version works because Warp can unroll that loop.

Warp can warn about potentially non-differentiable mutations inside dynamic loops during code generation. Enable debug diagnostics before the module is compiled:

```python
wp.config.log_level = wp.LOG_DEBUG
```

For this reproducer, Warp reports that `acc` is mutated inside a dynamic loop and that the operation may not be differentiable. This is a warning rather than an error because some dynamic loops produce correct gradients when the requested adjoints do not depend on values that would need to be replayed.

Specializing the kernel for a fixed loop count is a valid workaround. Since this loop only performs addition, another documented workaround is to move the dynamic loop into a separate `@wp.func`. Warp replays the function call during the backward pass, which restores the expected gradient for this example.

Support for replaying dynamic loops directly is already tracked in [#729](https://github.com/NVIDIA/warp/issues/729). We will use that issue to explore which parts of this limitation can be lifted safely. Since this report is another manifestation of the same documented limitation, I am closing it as a duplicate of #729.

**Comment by Ramzee-S** [comment]
Thanks for the clear mechanism — understood on the duplicate-of-#729 close, and apologies for missing the Dynamic Loops section before filing. Two verification notes from our side (warp 1.15.0), in case they're useful:

1. The `@wp.func` workaround checks out: wrapping the dynamic loop in a function restores the correct gradient for the nonlinear case too (`out = acc * b` with `acc` returned from a func → d(out)/dx = 12.0 as expected). That solved a second site in our codebase without restructuring.
2. A nuance worth a sentence in the guide: the stale value during replay is the *initializer*, not necessarily zero. Initializing `acc = float(5.0)` in the reproducer yields d(out)/dx = 10.0 (= 2 × 5) instead of the correct 22.0 — a plausible-looking wrong gradient rather than a conspicuous zero, which makes the failure mode harder to notice downstream of the documented behavior.

We've also adopted the `LOG_DEBUG` codegen warning as a CI check on our differentiable kernels. Watching #729.
